File: Java/Validate_Binary_Search_Tree.java
class Validate_Binary_Search_Tree{
	public boolean isValidBST(TreeNode root) {
        boolean a = isValid(root, Integer.MIN_VALUE, Integer.MAX_VALUE);
        boolean b = isValid(root,false,false,Integer.MAX_VALUE,Integer.MIN_VALUE);
        boolean c=  isValid2(root,Integer.MIN_VALUE,Integer.MAX_VALUE);
        return c;
    }
    private boolean isValid(TreeNode root, int rightMin, int leftMax){
    	if(root == null)
        	return true;
        if(root.val <= rightMin || root.val >= leftMax)
        	return false;
        return isValidBST(root.left,rightMin,root.val) && isValidBST(root.right,root.val,leftMax);
    }
    private boolean isValid(TreeNode root, boolean left, boolean right, int rightMin, int leftMax){
    	if(root == null)
        	return true;
        if(left && root.val >= leftMax)
        	return false;
        if(right && root.val <= rightMin)
        	return false;
        return isValid(root.left,true,right,rightMin,root.val) && isValid(root.right,left,true,root.val,leftMax);

    }
    private boolean isValid2(TreeNode root,int rightMin, int leftMax){
    	if(root == null)
        	return true;
        if(root.left != null && (root.left.val >= root.val || root.left.val <= rightMin))
        	return false;
        if(root.right != null && (root.right.val <= root.val || root.right.val >= leftMax))
        	return false;
        return isValid2(root.left,rightMin,root.val) && isValid2(root.right,root.val,leftMax);
    }_
}
class TreeNode {
    int val;
	TreeNode left;
	TreeNode right;
	TreeNode(int x) { val = x; }
}

File: Python/Validate Binary Search Tree.py
# Given a binary tree, determine if it is a valid binary search tree (BST).

# Assume a BST is defined as follows:

# The left subtree of a node contains only nodes with keys less than the node's key.
# The right subtree of a node contains only nodes with keys greater than the node's key.
# Both the left and right subtrees must also be binary search trees.

# Definition for a  binary tree node
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    # @param root, a tree node
    # @return a boolean
	def isValidBST(self, root):
		return self.verifyBST(root,False,False,0,0)
	def verifyBST(self,root,left,right,lMax,rMin):
		if root == None:
			return True
		if left and root.val >= lMax:
			return False
		if right and root.val <=rMin:
			return False
		leftValid = self.verifyBST(root.left,True,right,root.val,rMin)
		rightValid = self.verifyBST(root.right,left,True,lMax,root.val)
		return leftValid and rightValid